jaxopt/_src/projection.py

```diff
@@ -136,6 +136,60 @@ def projection_simplex(x: jnp.ndarray, value: float = 1.0) -> jnp.ndarray:
   return value * _projection_unit_simplex(x / value)
 
 
+@partial(jax.custom_jvp, nondiff_argnums=(1, 2, 3))
+def projection_simplex_k_sparse(x: jnp.ndarray, k: int, value: float = 1.0,
+                                use_approx_max_k: bool = False) -> jnp.ndarray:
+  r"""Create a k-sparse array (by keeping the top-k values and zeroing out others) and then project it onto a simplex.
+
+  By default, the projection is onto the probability simplex.
+
+  Args:
+    x: vector to project, an array of shape (n,).
+    k: the highest k values to keep
+    value: value p should sum to (default: 1.0).
+    use_approx_max_k: when set to True, use `jax.lax.approx_max_k` to return
+      max k values and their indices in an approximate manner (default: False).
+  Returns:
+    projected vector, an array with the same shape as ``x``.
+  """
+  if value is None:
+    value = 1.0
+  x /= value
+
+  # Top k values (in an decreasing order) and their corresponding indices
+  if use_approx_max_k:
+    top_k_values, top_k_indices = jax.lax.approx_max_k(x, k)
+  else:
+    top_k_values, top_k_indices = jax.lax.top_k(x, k)
+
+  # Projection the sorted top k values onto the unit simplex
+  cumsum_top_k_values = jnp.cumsum(top_k_values)
+  ind = jnp.arange(k) + 1
+  cond = 1 / ind + (top_k_values - cumsum_top_k_values / ind) > 0
+  idx = jnp.count_nonzero(cond)
+  top_k_simplex_projection = jax.nn.relu(
+      1 / idx + (x[top_k_indices] - cumsum_top_k_values[idx - 1] / idx))
+
+  # Put the projection of sorted top k values to their original indices;
+  # set all other indices zero.
+  k_sparse_simplex_projection = jnp.sum(
+      top_k_simplex_projection[ :, jnp.newaxis] * jax.nn.one_hot(
+          top_k_indices, len(x), dtype=x.dtype), axis=0)
+  return  k_sparse_simplex_projection
```

Comment:
Please create `_project_unit_sparse_simplex` the same way I did `_project_unit_simplex`. Currently your handling of `value` isn't correct (both in the forward and backward passes)